File: sdk/python/src/plinth/_http.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Internal HTTP wrapper around ``httpx`` with auth and error mapping.

This module is **not** part of the public API. It exists so that
``workspace.py``, ``tools.py``, and ``client.py`` can speak to the
backing services without each one re-implementing auth headers and the
``{"error": {...}}`` envelope unwrap.
"""

from __future__ import annotations

import logging
from typing import Any

import httpx

from .exceptions import (
    BranchNotFound,
    ChannelNotFound,
    CostCapExceeded,
    FileNotFound,
    InvalidArguments,
    InvalidToken,
    InvalidWorkflowStep,
    KeyNotFound,
    LeaseConflict,
    LeaseNotHeld,
    LockConflict,
    LockNotFound,
    LockNotHeld,
    MessageNotFound,
    PlinthError,
    RateLimited,
    SchemaViolation,
    SnapshotNotFound,
    TokenExpired,
    TokenRevoked,
    ToolInvocationError,
    ToolNotFound,
    TransactionInvalidStatus,
    TransactionNotFound,
    Unauthorized,
    WorkerNotFound,
    WorkflowNotFound,
    WorkflowStepNotFound,
    WorkspaceNotFound,
)

# Map error codes from the Plinth error envelope to typed exceptions.
# When a 404 arrives without an explicit code, the caller hints the
# expected resource via ``not_found_class`` on the request method.
_CODE_TO_EXCEPTION: dict[str, type[PlinthError]] = {
    "WORKSPACE_NOT_FOUND": WorkspaceNotFound,
    "KEY_NOT_FOUND": KeyNotFound,
    "FILE_NOT_FOUND": FileNotFound,
    "SNAPSHOT_NOT_FOUND": SnapshotNotFound,
    "BRANCH_NOT_FOUND": BranchNotFound,
    "TOOL_NOT_FOUND": ToolNotFound,
    "CHANNEL_NOT_FOUND": ChannelNotFound,
    "MESSAGE_NOT_FOUND": MessageNotFound,
    "WORKFLOW_NOT_FOUND": WorkflowNotFound,
    "WORKFLOW_STEP_NOT_FOUND": WorkflowStepNotFound,
    "INVALID_WORKFLOW_STEP": InvalidWorkflowStep,
    "TRANSACTION_NOT_FOUND": TransactionNotFound,
    "TRANSACTION_INVALID_STATUS": TransactionInvalidStatus,
    "TRANSACTION_RENDER_ERROR": InvalidArguments,
    "SCHEMA_VIOLATION": SchemaViolation,
    "INVALID_ARGUMENTS": InvalidArguments,
    "UNAUTHORIZED": Unauthorized,
    "INVALID_TOKEN": InvalidToken,
    "TOKEN_EXPIRED": TokenExpired,
    "TOKEN_REVOKED": TokenRevoked,
    "RATE_LIMITED": RateLimited,
    "COST_CAP_EXCEEDED": CostCapExceeded,
    "TOOL_INVOCATION_FAILED": ToolInvocationError,
    "LEASE_CONFLICT": LeaseConflict,
    "LEASE_NOT_HELD": LeaseNotHeld,
    "WORKER_NOT_FOUND": WorkerNotFound,
    # v0.6 — generic resource locks. The workspace service emits
    # ``LOCK_HELD`` on contention; the SDK exposes it as :class:`LockConflict`
    # so user code reads naturally (``except LockConflict:``). The
    # ``LOCK_CONFLICT`` alias is accepted for symmetry with future services
    # that may surface the spec's preferred code directly.
    "LOCK_HELD": LockConflict,
    "LOCK_CONFLICT": LockConflict,
    "LOCK_NOT_HELD": LockNotHeld,
    "LOCK_NOT_FOUND": LockNotFound,
}

_STATUS_TO_EXCEPTION: dict[int, type[PlinthError]] = {
    400: InvalidArguments,
    401: Unauthorized,
    429: RateLimited,
}


_logger = logging.getLogger("plinth.sdk.http")


class HTTPClient:
    """Thin wrapper around ``httpx.Client`` that adds Plinth auth & errors.

    A single ``HTTPClient`` is bound to one primary base URL (workspace or
    gateway). The ``Plinth`` facade owns one wrapper per service.

    v1.0 — multi-region. Pass ``fallback_urls={region_id: url, ...}`` and
    the client will try fallbacks in iteration order on:

    * connection errors (``httpx.ConnectError``, ``httpx.ConnectTimeout``,
      ``httpx.ReadTimeout``)
    * 5xx responses (the primary is up but degraded)
    * 503 responses
    * 421 / 409 responses carrying an ``X-Plinth-Primary-Region`` and/or
      ``X-Plinth-Primary-URL`` header — when the named region exists in
      ``fallback_urls`` (or the URL header points at a known peer) we
      retry there directly. This is the read-replica redirect path.
      ``X-Plinth-Primary-URL`` wins over the region lookup when both are
      present, so a replica with no peer wired into the SDK can still
      hand back the canonical primary location.

    The redirect retry is **bounded**: each unique base URL is tried at
    most once per request (tracked in an ``attempted`` set) so a
    misconfigured pair of replicas can never produce an infinite loop.

    Fallback ordering is *deterministic*: ``fallback_urls`` is a dict so
    iteration order matches insertion order (Python 3.7+ guarantee).
    Tests that rely on it pass a regular ``dict`` and get the same
    behaviour the operator would in production.
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float = 30.0,
        *,
        transport: httpx.BaseTransport | None = None,
        fallback_urls: dict[str, str] | None = None,
        primary_region: str | None = None,
    ) -> None:
        self._headers = {
            "Authorization": f"Bearer {api_key}",
            "User-Agent": "plinth-python-sdk/0.1.0",
        }
        self._timeout = timeout
        self._transport = transport
        # The primary URL is always tried first.
        self._primary_url = base_url.rstrip("/")
        self._primary_region = primary_region
        # Deterministic ordering; dict preserves insertion order.
        self._fallback_urls: dict[str, str] = {
            region: url.rstrip("/")
            for region, url in (fallback_urls or {}).items()
        }
        # ``transport`` is wired by tests using ``respx.MockTransport``;
        # in production callers leave it ``None`` and httpx picks the
        # default async-capable transport.
        self._client = httpx.Client(
            base_url=self._primary_url,
            headers=self._headers,
            timeout=timeout,
            transport=transport,
        )
        # Per-fallback client, lazily built on first use.
        self._fallback_clients: dict[str, httpx.Client] = {}
        # Track which URL is the current primary. Failover updates this
        # so subsequent requests don't re-try the dead primary first.
        self._current_url = self._primary_url

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Close the underlying connection pool."""
        self._client.close()
        for client in self._fallback_clients.values():
            try:
                client.close()
            except Exception:  # pragma: no cover - best-effort
                pass

    def __enter__(self) -> HTTPClient:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    # ------------------------------------------------------------------
    # Verb helpers
    # ------------------------------------------------------------------

    def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> httpx.Response:
        """Issue a GET request and return the raw response."""
        return self._request(
            "GET",
            path,
            params=params,
            not_found_class=not_found_class,
        )

    def post(
        self,
        path: str,
        *,
        json: Any | None = None,
        content: bytes | None = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> httpx.Response:
        """Issue a POST request and return the raw response."""
        return self._request(
            "POST",
            path,
            json=json,
            content=content,
            params=params,
            extra_headers=headers,
            not_found_class=not_found_class,
        )

    def put(
        self,
        path: str,
        *,
        json: Any | None = None,
        content: bytes | None = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> httpx.Response:
        """Issue a PUT request and return the raw response."""
        return self._request(
            "PUT",
            path,
            json=json,
            content=content,
            params=params,
            extra_headers=headers,
            not_found_class=not_found_class,
        )

    def delete(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> httpx.Response:
        """Issue a DELETE request and return the raw response."""
        return self._request(
            "DELETE",
            path,
            params=params,
            not_found_class=not_found_class,
        )

    def get_json(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> Any:
        """Convenience: GET and return the JSON-decoded body."""
        return self.get(path, params=params, not_found_class=not_found_class).json()

    # ------------------------------------------------------------------
    # Multi-region failover
    # ------------------------------------------------------------------

    def _client_for(self, base_url: str) -> httpx.Client:
        """Return (or lazily build) an ``httpx.Client`` bound to ``base_url``."""

        if base_url == self._primary_url:
            return self._client
        existing = self._fallback_clients.get(base_url)
        if existing is not None:
            return existing
        new_client = httpx.Client(
            base_url=base_url,
            headers=self._headers,
            timeout=self._timeout,
            transport=self._transport,
        )
        self._fallback_clients[base_url] = new_client
        return new_client

    def _candidates_in_order(self) -> list[tuple[str, str]]:
        """Return ``[(region_id, url), ...]`` to try, in priority order.

        The primary always goes first (under its declared region id when
        ``primary_region`` is configured, else as ``"<primary>"``). The
        rest are the fallbacks in insertion order.
        """

        seq: list[tuple[str, str]] = [(self._primary_region or "<primary>", self._primary_url)]
        for region, url in self._fallback_urls.items():
            seq.append((region, url))
        return seq

    def _resolve_redirect_target(self, response: httpx.Response) -> str | None:
        """Pick the redirect URL from ``X-Plinth-Primary-*`` headers.

        Resolution order:

        1. ``X-Plinth-Primary-URL`` — explicit URL the server told us to
           use. We trust it iff its origin already appears in our known
           candidate set (primary URL or any configured fallback URL).
           Trusting a header URL blindly would let a malicious response
           steer the SDK at an attacker-controlled host.
        2. ``X-Plinth-Primary-Region`` looked up against ``fallback_urls``.
        3. ``X-Plinth-Primary-Region`` matching ``primary_region`` →
           the configured primary URL itself.

        Returns ``None`` when no trusted target can be derived.
        """

        url_hint = (response.headers.get("X-Plinth-Primary-URL") or "").strip()
        if url_hint:
            normalized = url_hint.rstrip("/")
            known = {self._primary_url, *self._fallback_urls.values()}
            if normalized in known:
                return normalized

        target_region = (
            response.headers.get("X-Plinth-Primary-Region") or ""
        ).strip()
        if not target_region:
            return None
        target_url = self._fallback_urls.get(target_region)
        if target_url is not None:
            return target_url
        if self._primary_region == target_region:
            return self._primary_url
        return None

    def _request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        content: bytes | None = None,
        params: dict[str, Any] | None = None,
        extra_headers: dict[str, str] | None = None,
        not_found_class: type[PlinthError] | None = None,
    ) -> httpx.Response:
        """Issue a single request, with multi-region failover.

        Failover sequence:

        1. Try the primary URL.
        2. On connection error / 503 / 5xx: try each fallback in order.
        3. On 421 (Misdirected Request) or 409 + replica-redirect headers:
           extract the target URL via ``X-Plinth-Primary-URL`` (preferred)
           or look up the region named in ``X-Plinth-Primary-Region``
           against ``fallback_urls``, and retry there directly. The
           previously-primary URL is *not* re-tried — the server told us
           it's a replica.

        Each unique base URL is tried at most once (tracked via the
        ``attempted`` set) so a misconfigured pair of replicas can't
        loop forever.

        Errors from the *last* attempt are re-raised; intermediate
        failures emit a structured warning so operators see them.
        """

        candidates = self._candidates_in_order()
        if not candidates:  # pragma: no cover - constructor enforces ≥1
            raise PlinthError("no candidate URLs configured", code="INTERNAL")

        last_exc: PlinthError | None = None
        attempted: set[str] = set()
        # We may add a 409-redirect target dynamically; track the working
        # queue separately from the static candidate list.
        queue: list[tuple[str, str]] = list(candidates)
        idx = 0
        while idx < len(queue):
            region, base_url = queue[idx]
            idx += 1
            if base_url in attempted:
                continue
            attempted.add(base_url)
            client = self._client_for(base_url)
            try:
                response = self._send_one(
                    client,
                    method,
                    path,
                    json=json,
                    content=content,
                    params=params,
                    extra_headers=extra_headers,
                )
            except httpx.HTTPError as exc:
                last_exc = PlinthError(
                    f"connection error to {base_url}: {exc}",
                    code="CONNECTION_ERROR",
                )
                _logger.warning(
                    "plinth.sdk.failover: connection_error region=%s url=%s err=%s",
                    region,
                    base_url,
                    exc,
                )
                continue

            # Handle replica-redirect responses. The server emits 421
            # (Misdirected Request, RFC 7540) for read-replica writes,
            # alongside ``X-Plinth-Primary-URL`` (preferred) and
            # ``X-Plinth-Primary-Region`` headers. We also accept 409
            # for backwards compatibility with pre-spec implementations.
            if response.status_code in (421, 409) and (
                response.headers.get("X-Plinth-Primary-Region")
                or response.headers.get("X-Plinth-Primary-URL")
            ):
                target_url = self._resolve_redirect_target(response)
                target_region = (
                    response.headers.get("X-Plinth-Primary-Region", "").strip()
                    or "<primary>"
                )
                if target_url is not None and target_url not in attempted:
                    _logger.warning(
                        "plinth.sdk.failover: replica_redirect region=%s url=%s -> %s",
                        region,
                        base_url,
                        target_url,
                    )
                    # Insert the redirect target as the next item. Each
                    # URL is attempted at most once per request via the
                    # ``attempted`` set above, so loops are impossible.
                    queue.insert(idx, (target_region, target_url))
                    continue
                # No fallback for the named region: surface the error.
                self._raise_for_status(response, not_found_class=not_found_class)
                return response

            # Retry candidates on 503 / 5xx that smell like the local
            # instance being unreachable / overloaded.
            if response.status_code in (502, 503, 504) and idx < len(queue):
                last_exc = PlinthError(
                    f"upstream {response.status_code} from {base_url}",
                    code="UPSTREAM_DEGRADED",
                    response=response,
                )
                _logger.warning(
                    "plinth.sdk.failover: upstream_5xx region=%s url=%s status=%s",
                    region,
                    base_url,
                    response.status_code,
                )
                continue

            self._raise_for_status(response, not_found_class=not_found_class)
            return response

        # Every candidate failed. Re-raise the last error.
        if last_exc is not None:
            raise last_exc
        raise PlinthError("no candidate URLs succeeded", code="INTERNAL")

    def _send_one(
        self,
        client: httpx.Client,
        method: str,
        path: str,
        *,
        json: Any | None,
        content: bytes | None,
        params: dict[str, Any] | None,
        extra_headers: dict[str, str] | None,
    ) -> httpx.Response:
        """Issue the actual HTTP request through ``client``.

        Centralised so the failover loop has one entry point per attempt.
        """

        kwargs: dict[str, Any] = {
            "params": _clean_params(params),
        }
        if extra_headers is not None:
            kwargs["headers"] = extra_headers
        if method == "GET":
            return client.get(path, **kwargs)
        if method == "DELETE":
            return client.delete(path, **kwargs)
        # POST/PUT/PATCH: include json/content payloads.
        kwargs["json"] = json
        kwargs["content"] = content
        return client.request(method, path, **kwargs)

    # ------------------------------------------------------------------
    # Error mapping
    # ------------------------------------------------------------------

    def _raise_for_status(
        self,
        response: httpx.Response,
        *,
        not_found_class: type[PlinthError] | None = None,
    ) -> None:
        """Map any non-2xx response to a typed Plinth exception."""
        if response.is_success:
            return

        envelope = _safe_json(response)
        error_obj = (envelope or {}).get("error", {}) if isinstance(envelope, dict) else {}
        code = error_obj.get("code")
        message = error_obj.get("message") or response.text or response.reason_phrase
        details = error_obj.get("details") or {}

        # 1. Prefer the explicit error code from the response envelope.
        exc_class: type[PlinthError] | None = _CODE_TO_EXCEPTION.get(code) if code else None

        # 2. Fall back to the status-code map.
        if exc_class is None:
            exc_class = _STATUS_TO_EXCEPTION.get(response.status_code)

        # 3. For 404s, prefer the resource-specific class hinted by the
        #    caller (e.g. ``KeyNotFound`` for the KV endpoints).
        if exc_class is None and response.status_code == 404:
            exc_class = not_found_class or PlinthError

        # 4. Final catch-all: 5xx and anything else.
        if exc_class is None:
            exc_class = PlinthError

        # ``RateLimited`` (and its ``CostCapExceeded`` subclass) carry
        # extra retry metadata pulled from the response. We compute it
        # here so callers can sleep/back-off without re-reading headers.
        if response.status_code == 429 or issubclass(exc_class, RateLimited):
            retry_after = _parse_retry_after(response, details)
            reason = details.get("limit_type", "") if isinstance(details, dict) else ""
            current = details.get("current") if isinstance(details, dict) else None
            limit = details.get("limit") if isinstance(details, dict) else None
            raise exc_class(
                message,
                code=code,
                details=details,
                response=response,
                retry_after=retry_after,
                reason=reason or "",
                current=current,
                limit=limit,
            )

        # ``LockConflict`` (v0.6) — surface ``current_holder`` and the
        # server's back-off hint directly so callers don't have to dig
        # through ``e.details``.
        if issubclass(exc_class, LockConflict):
            current_holder = (
                details.get("current_holder")
                if isinstance(details, dict)
                else None
            )
            retry_after_seconds = (
                details.get("retry_after_seconds")
                if isinstance(details, dict)
                else None
            )
            raise exc_class(
                message,
                code=code,
                details=details,
                response=response,
                current_holder=current_holder,
                retry_after_seconds=retry_after_seconds,
            )

        # ``SchemaViolation`` (v0.5) — surface the validator errors and the
        # DLQ message ID directly on the exception so callers don't need to
        # reach into ``e.details``.
        if issubclass(exc_class, SchemaViolation):
            errors = (
                details.get("errors")
                if isinstance(details, dict) and isinstance(details.get("errors"), list)
                else []
            )
            dlq = (
                details.get("deadletter_msg_id")
                if isinstance(details, dict)
                else None
            )
            channel = details.get("channel") if isinstance(details, dict) else None
            raise exc_class(
                message,
                code=code,
                details=details,
                response=response,
                errors=errors,
                deadletter_msg_id=dlq,
                channel=channel,
            )

        raise exc_class(message, code=code, details=details, response=response)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _safe_json(response: httpx.Response) -> Any | None:
    """Best-effort JSON decode; returns ``None`` for non-JSON bodies."""
    try:
        return response.json()
    except (ValueError, UnicodeDecodeError):
        return None


def _clean_params(params: dict[str, Any] | None) -> dict[str, Any] | None:
    """Strip ``None`` values so they don't end up as ``?key=None``."""
    if not params:
        return None
    return {k: v for k, v in params.items() if v is not None}


def _parse_retry_after(
    response: httpx.Response,
    details: dict[str, Any] | None,
) -> float:
    """Extract a ``retry_after`` (seconds) from a 429 response.

    Prefers the structured ``details.retry_after_seconds`` from the error
    envelope, then falls back to the ``Retry-After`` HTTP header. Returns
    ``0.0`` when neither is parsable.
    """
    if isinstance(details, dict):
        raw = details.get("retry_after_seconds")
        if raw is not None:
            try:
                return float(raw)
            except (TypeError, ValueError):
                pass

    header = response.headers.get("Retry-After") or response.headers.get("retry-after")
    if header:
        try:
            return float(header)
        except (TypeError, ValueError):
            return 0.0
    return 0.0


__all__ = ["HTTPClient"]

File: sdk/go/plinth/http.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// HTTPClient is the internal request helper bound to a single Plinth
// service base URL. The Plinth facade owns one per backing service
// (workspace, gateway, identity).
//
// Public API surface intentionally uses the standard library: any
// *http.Client passed in via Config.HTTPClient is used as-is, so
// callers can wire up custom transports, retries, or instrumentation.
type HTTPClient struct {
	baseURL    string
	apiKey     string
	userAgent  string
	httpClient *http.Client
}

// NewHTTPClient returns a Plinth HTTPClient bound to baseURL.
//
// The userAgent and apiKey are attached to every outgoing request.
// httpClient must be non-nil — callers should pass a standard
// *http.Client with a reasonable timeout (the Plinth facade defaults
// to 30s when none is supplied).
func NewHTTPClient(httpClient *http.Client, baseURL, apiKey, userAgent string) *HTTPClient {
	return &HTTPClient{
		baseURL:    strings.TrimRight(baseURL, "/"),
		apiKey:     apiKey,
		userAgent:  userAgent,
		httpClient: httpClient,
	}
}

// BaseURL returns the base URL the client was constructed with (no
// trailing slash). Useful for diagnostic logging.
func (c *HTTPClient) BaseURL() string { return c.baseURL }

// requestOptions bundles the optional flags accepted by the verb
// helpers. Kept private — callers use Get/Post/Put/Delete and the JSON
// variants instead.
type requestOptions struct {
	body        any              // marshaled to JSON when set
	rawBody     []byte           // sent as-is when set (overrides body)
	contentType string           // overrides default Content-Type
	query       url.Values       // attached as ?key=value
	headers     map[string]string // extra request headers
	notFoundCode string          // override "NOT_FOUND" sentinel for 404s
}

// RequestOption mutates a requestOptions before the request is built.
// This package-private functional-options pattern keeps the verb method
// signatures small while still allowing every flag the Python SDK
// exposes via kwargs.
type RequestOption func(*requestOptions)

// WithJSON attaches a JSON body to the request.
func WithJSON(body any) RequestOption {
	return func(o *requestOptions) { o.body = body }
}

// WithBody attaches a raw byte body. Use for file uploads where the
// caller has already produced bytes and a content type.
func WithBody(b []byte, contentType string) RequestOption {
	return func(o *requestOptions) {
		o.rawBody = b
		if contentType != "" {
			o.contentType = contentType
		}
	}
}

// WithQuery attaches query parameters. Nil values are dropped.
func WithQuery(q url.Values) RequestOption {
	return func(o *requestOptions) { o.query = q }
}

// WithHeader attaches a single extra HTTP header to the request.
// Multiple WithHeader options compose.
func WithHeader(key, value string) RequestOption {
	return func(o *requestOptions) {
		if o.headers == nil {
			o.headers = make(map[string]string)
		}
		o.headers[key] = value
	}
}

// WithNotFoundCode hints the resource-specific error code to attach
// when the server returns a 404 without a code in the envelope. This
// matches the Python SDK's not_found_class kwarg.
func WithNotFoundCode(code string) RequestOption {
	return func(o *requestOptions) { o.notFoundCode = code }
}

// Get issues a GET request and returns the raw response body bytes.
//
// Non-2xx responses are mapped to *PlinthError before this returns.
func (c *HTTPClient) Get(ctx context.Context, path string, opts ...RequestOption) ([]byte, error) {
	return c.do(ctx, http.MethodGet, path, opts...)
}

// GetJSON issues a GET request and decodes the JSON body into out.
// out should be a pointer; passing a non-pointer is a programmer error
// and will surface as a json.Unmarshal failure.
func (c *HTTPClient) GetJSON(ctx context.Context, path string, out any, opts ...RequestOption) error {
	body, err := c.Get(ctx, path, opts...)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	return decodeJSON(body, out)
}

// Post issues a POST request and returns the raw response body bytes.
func (c *HTTPClient) Post(ctx context.Context, path string, opts ...RequestOption) ([]byte, error) {
	return c.do(ctx, http.MethodPost, path, opts...)
}

// PostJSON issues a POST request and decodes the JSON body into out.
func (c *HTTPClient) PostJSON(ctx context.Context, path string, out any, opts ...RequestOption) error {
	body, err := c.Post(ctx, path, opts...)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	return decodeJSON(body, out)
}

// Put issues a PUT request and returns the raw response body bytes.
func (c *HTTPClient) Put(ctx context.Context, path string, opts ...RequestOption) ([]byte, error) {
	return c.do(ctx, http.MethodPut, path, opts...)
}

// PutJSON issues a PUT request and decodes the JSON body into out.
func (c *HTTPClient) PutJSON(ctx context.Context, path string, out any, opts ...RequestOption) error {
	body, err := c.Put(ctx, path, opts...)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	return decodeJSON(body, out)
}

// Patch issues a PATCH request and returns the raw response body bytes.
func (c *HTTPClient) Patch(ctx context.Context, path string, opts ...RequestOption) ([]byte, error) {
	return c.do(ctx, http.MethodPatch, path, opts...)
}

// PatchJSON issues a PATCH request and decodes the JSON body into out.
func (c *HTTPClient) PatchJSON(ctx context.Context, path string, out any, opts ...RequestOption) error {
	body, err := c.Patch(ctx, path, opts...)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	return decodeJSON(body, out)
}

// Delete issues a DELETE request and discards the response body.
func (c *HTTPClient) Delete(ctx context.Context, path string, opts ...RequestOption) error {
	_, err := c.do(ctx, http.MethodDelete, path, opts...)
	return err
}

// do is the single request entry point used by every verb helper. It
// builds the *http.Request, sends it, decodes the standard error
// envelope on non-2xx, and returns the raw body on success.
func (c *HTTPClient) do(ctx context.Context, method, path string, opts ...RequestOption) ([]byte, error) {
	o := &requestOptions{}
	for _, opt := range opts {
		opt(o)
	}

	endpoint, err := c.buildURL(path, o.query)
	if err != nil {
		return nil, &PlinthError{Code: ErrInternal.Code, Message: err.Error(), Cause: err}
	}

	var bodyReader io.Reader
	contentType := ""
	switch {
	case o.rawBody != nil:
		bodyReader = bytes.NewReader(o.rawBody)
		contentType = o.contentType
		if contentType == "" {
			contentType = "application/octet-stream"
		}
	case o.body != nil:
		buf, marshalErr := json.Marshal(o.body)
		if marshalErr != nil {
			return nil, &PlinthError{
				Code:    ErrInternal.Code,
				Message: fmt.Sprintf("marshal request body: %v", marshalErr),
				Cause:   marshalErr,
			}
		}
		bodyReader = bytes.NewReader(buf)
		contentType = "application/json"
	}

	req, err := http.NewRequestWithContext(ctx, method, endpoint, bodyReader)
	if err != nil {
		return nil, &PlinthError{Code: ErrInternal.Code, Message: err.Error(), Cause: err}
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("Accept", "application/json, application/octet-stream")
	if c.userAgent != "" {
		req.Header.Set("User-Agent", c.userAgent)
	}
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	for k, v := range o.headers {
		req.Header.Set(k, v)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		// net errors include context cancellation, DNS failure, etc.
		// Wrap so callers can errors.Is(err, plinth.ErrConnectionError).
		return nil, &PlinthError{
			Code:    ErrConnectionError.Code,
			Message: fmt.Sprintf("connection error to %s: %v", endpoint, err),
			Cause:   err,
		}
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &PlinthError{
			Code:    ErrConnectionError.Code,
			Message: fmt.Sprintf("read response body from %s: %v", endpoint, err),
			Cause:   err,
		}
	}

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return body, nil
	}

	return nil, c.errorForResponse(resp, body, o.notFoundCode)
}

// buildURL joins the request path onto baseURL and appends any non-nil
// query parameters. Nil/empty values in the url.Values are skipped to
// avoid emitting "?key=" dangling fragments — matches the Python SDK's
// _clean_params helper.
func (c *HTTPClient) buildURL(path string, query url.Values) (string, error) {
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	full := c.baseURL + path
	if len(query) == 0 {
		return full, nil
	}
	cleaned := url.Values{}
	for k, vs := range query {
		for _, v := range vs {
			if v == "" {
				continue
			}
			cleaned.Add(k, v)
		}
	}
	if len(cleaned) == 0 {
		return full, nil
	}
	separator := "?"
	if strings.Contains(full, "?") {
		separator = "&"
	}
	return full + separator + cleaned.Encode(), nil
}

// errorForResponse decodes the standard {"error": {...}} envelope and
// returns a *PlinthError with the right Code/Sentinel attached.
func (c *HTTPClient) errorForResponse(resp *http.Response, body []byte, fallbackNotFoundCode string) error {
	envelope := struct {
		Error struct {
			Code    string         `json:"code"`
			Message string         `json:"message"`
			Details map[string]any `json:"details"`
		} `json:"error"`
	}{}
	_ = json.Unmarshal(body, &envelope) // best-effort

	code := envelope.Error.Code
	message := envelope.Error.Message
	details := envelope.Error.Details

	// If the server didn't include a code, fall back first to the
	// caller-supplied 404 hint, then to the status-code map.
	if code == "" {
		if resp.StatusCode == http.StatusNotFound && fallbackNotFoundCode != "" {
			code = fallbackNotFoundCode
		} else if mapped, ok := statusToCode[resp.StatusCode]; ok {
			code = mapped
		} else {
			code = ErrInternal.Code
		}
	}

	if message == "" {
		if len(body) > 0 {
			message = strings.TrimSpace(string(body))
		} else {
			message = http.StatusText(resp.StatusCode)
		}
	}

	pe := &PlinthError{
		Code:       code,
		Message:    message,
		Details:    details,
		StatusCode: resp.StatusCode,
		Body:       body,
	}

	// Rate-limit responses carry retry hints in the envelope and/or
	// the Retry-After header. Surface them on the PlinthError so
	// callers can sleep without re-reading the response.
	if resp.StatusCode == http.StatusTooManyRequests || code == "RATE_LIMITED" || code == "COST_CAP_EXCEEDED" {
		pe.RetryAfter = parseRetryAfter(resp, details)
		if lt, ok := details["limit_type"].(string); ok {
			pe.LimitType = lt
		}
	}

	// errors.Is dispatch happens automatically: PlinthError.Is matches
	// on Code, so the returned *PlinthError satisfies the matching
	// sentinel without us having to wrap a Cause. We do consult the
	// table to canonicalise legacy aliases (e.g. server-side
	// "LOCK_HELD" → SDK-side "LOCK_CONFLICT") so callers can rely on
	// a single code per concept.
	if sentinel, ok := codeToSentinel[code]; ok && sentinel.Code != code {
		pe.Code = sentinel.Code
	}
	return pe
}

// parseRetryAfter extracts a retry hint (seconds) from the response.
// Prefers details.retry_after_seconds, falls back to the Retry-After
// header. Returns 0 when neither is present or parsable.
func parseRetryAfter(resp *http.Response, details map[string]any) float64 {
	if details != nil {
		switch raw := details["retry_after_seconds"].(type) {
		case float64:
			return raw
		case int:
			return float64(raw)
		case int64:
			return float64(raw)
		case string:
			if v, err := strconv.ParseFloat(raw, 64); err == nil {
				return v
			}
		}
	}
	header := resp.Header.Get("Retry-After")
	if header == "" {
		return 0
	}
	if v, err := strconv.ParseFloat(header, 64); err == nil {
		return v
	}
	return 0
}

// decodeJSON wraps json.Unmarshal with a PlinthError on failure so
// every error returned by the SDK is uniformly typed.
func decodeJSON(body []byte, out any) error {
	if len(body) == 0 {
		return nil
	}
	if err := json.Unmarshal(body, out); err != nil {
		return &PlinthError{
			Code:    ErrInternal.Code,
			Message: fmt.Sprintf("decode JSON: %v", err),
			Cause:   err,
			Body:    body,
		}
	}
	return nil
}

// EncodePathSegment percent-encodes a single path segment. Mirrors
// the Python SDK's _ek helper for KV keys and similar opaque values.
//
// Use this — not url.PathEscape — when the caller-provided string may
// contain "/" but should be treated as a single segment.
func EncodePathSegment(s string) string {
	return url.PathEscape(s)
}

// EncodeFilePath percent-encodes a file path, preserving "/" as the
// segment separator. Mirrors the Python SDK's _ep helper.
func EncodeFilePath(p string) string {
	parts := strings.Split(strings.TrimLeft(p, "/"), "/")
	for i, part := range parts {
		parts[i] = url.PathEscape(part)
	}
	return strings.Join(parts, "/")
}

// EncodeLockName percent-encodes a lock name while preserving both "/"
// and ":" so the workspace's {name:path} route can consume the whole
// thing as one segment.
func EncodeLockName(name string) string {
	parts := strings.Split(strings.TrimLeft(name, "/"), "/")
	for i, part := range parts {
		// Re-allow ":" since it's a valid sub-delimiter inside a path
		// component per RFC 3986.
		parts[i] = strings.ReplaceAll(url.PathEscape(part), "%3A", ":")
	}
	return strings.Join(parts, "/")
}

// IsCode reports whether err carries the given Plinth error code.
// Convenience helper for code-driven branching without writing a full
// errors.As block.
func IsCode(err error, code string) bool {
	var pe *PlinthError
	if !errors.As(err, &pe) {
		return false
	}
	return pe.Code == code
}
